### PR TITLE
vulkan: simplify function calls

### PR DESCRIPTION
**Pull request by scribam**
Vulkan-Hpp provides enhanced versions of the commands and enhanced structure constructors which simplify the handling of array data.

### CODE REVIEW

File: core/rend/vulkan/oit/oit_pipeline.h [hunk 1]
```diff
@@ -162,45 +162,43 @@ class OITDescriptorSets
...
 
-			vk::DescriptorImageInfo imageInfo0;
 			if (poly.texture != nullptr)
 			{
-				imageInfo0 = vk::DescriptorImageInfo{ samplerManager->GetSampler(poly.tsp), ((Texture *)poly.texture)->GetReadOnlyImageView(),
-						vk::ImageLayout::eShaderReadOnlyOptimal };
-				writeDescriptorSets.emplace_back(perPolyDescSet, 0, 0, 1, vk::DescriptorType::eCombinedImageSampler, &imageInfo0, nullptr, nullptr);
+				vk::DescriptorImageInfo imageInfo0(samplerManager->GetSampler(poly.tsp), ((Texture *)poly.texture)->GetReadOnlyImageView(),
+						vk::ImageLayout::eShaderReadOnlyOptimal);
+				writeDescriptorSets.emplace_back(perPolyDescSet, 0, 0, vk::DescriptorType::eCombinedImageSampler, imageInfo0);
```
Comment (flyinghead) [inline core/rend/vulkan/oit/oit_pipeline.h:169]: I believe there is an issue here: imageInfo0 is used by reference in the emplace_back/constructor (vk::ArrayProxyNoTemporaries) but goes out of scope before `writeDescriptorSets` is used.

Same thing for imageInfo1, uniBufferInfo and lightBufferInfo below.

This is the reason I had these variables defined outside the `if`'s

File: core/rend/vulkan/oit/oit_pipeline.h [hunk 2]
```diff
@@ -162,45 +162,43 @@ class OITDescriptorSets
...
-			vk::DescriptorImageInfo imageInfo1;
+
 			if (poly.texture1 != nullptr)
 			{
-				imageInfo1 = vk::DescriptorImageInfo{ samplerManager->GetSampler(poly.tsp1), ((Texture *)poly.texture1)->GetReadOnlyImageView(),
-					vk::ImageLayout::eShaderReadOnlyOptimal };
-				writeDescriptorSets.emplace_back(perPolyDescSet, 1, 0, 1, vk::DescriptorType::eCombinedImageSampler, &imageInfo1, nullptr, nullptr);
+				vk::DescriptorImageInfo imageInfo1(samplerManager->GetSampler(poly.tsp1), ((Texture *)poly.texture1)->GetReadOnlyImageView(),
+					vk::ImageLayout::eShaderReadOnlyOptimal);
+				writeDescriptorSets.emplace_back(perPolyDescSet, 1, 0, vk::DescriptorType::eCombinedImageSampler, imageInfo1);
```
Comment (flyinghead) [inline core/rend/vulkan/oit/oit_pipeline.h:176]: imageInfo1 is used be reference but goes out of scope

File: core/rend/vulkan/oit/oit_pipeline.h [hunk 3]
```diff
@@ -162,45 +162,43 @@ class OITDescriptorSets
...
 
 				size = sizeof(N2LightModel) + align(sizeof(N2LightModel), uniformAlignment);
+				vk::DescriptorBufferInfo lightBufferInfo;
 				// light at index 0 is no light
 				if (poly.lightModel != nullptr)
 					lightBufferInfo = vk::DescriptorBufferInfo{ buffer, lightOffset + (poly.lightModel - pvrrc.lightModels.head() + 1) * size, sizeof(N2LightModel) };
 				else
 					lightBufferInfo = vk::DescriptorBufferInfo{ buffer, lightOffset, sizeof(N2LightModel) };
-				writeDescriptorSets.emplace_back(perPolyDescSet, 3, 0, 1, vk::DescriptorType::eUniformBuffer, nullptr, &lightBufferInfo, nullptr);
+				writeDescriptorSets.emplace_back(perPolyDescSet, 3, 0, vk::DescriptorType::eUniformBuffer, nullptr, lightBufferInfo);
```
Comment (flyinghead) [inline core/rend/vulkan/oit/oit_pipeline.h:193]: uniBufferInfo and lightBufferInfo are used by reference but go out of scope

File: core/rend/vulkan/pipeline.h [hunk 4]
```diff
@@ -86,38 +86,36 @@ class DescriptorSets
...
 			std::vector<vk::WriteDescriptorSet> writeDescriptorSets;
 
-			vk::DescriptorImageInfo imageInfo;
 			if (poly.texture != nullptr)
 			{
-				imageInfo = vk::DescriptorImageInfo(samplerManager->GetSampler(poly.tsp),
+				vk::DescriptorImageInfo imageInfo(samplerManager->GetSampler(poly.tsp),
 						((Texture *)poly.texture)->GetReadOnlyImageView(), vk::ImageLayout::eShaderReadOnlyOptimal);
-				writeDescriptorSets.emplace_back(perPolyDescSet, 0, 0, 1, vk::DescriptorType::eCombinedImageSampler, &imageInfo, nullptr, nullptr);
+				writeDescriptorSets.emplace_back(perPolyDescSet, 0, 0, vk::DescriptorType::eCombinedImageSampler, imageInfo);
```
Comment (flyinghead) [inline core/rend/vulkan/pipeline.h:93]: imageInfo is used by reference and goes out of scope

File: core/rend/vulkan/pipeline.h [hunk 5]
```diff
@@ -86,38 +86,36 @@ class DescriptorSets
...
 
 				size = sizeof(N2LightModel) + align(sizeof(N2LightModel), uniformAlignment);
+				vk::DescriptorBufferInfo lightBufferInfo;
 				// light at index 0 is no light
 				if (poly.lightModel != nullptr)
 					lightBufferInfo = vk::DescriptorBufferInfo{ buffer, lightOffset + (poly.lightModel - pvrrc.lightModels.head() + 1) * size, sizeof(N2LightModel) };
 				else
 					lightBufferInfo = vk::DescriptorBufferInfo{ buffer, lightOffset, sizeof(N2LightModel) };
-				writeDescriptorSets.emplace_back(perPolyDescSet, 3, 0, 1, vk::DescriptorType::eUniformBuffer, nullptr, &lightBufferInfo, nullptr);
+				writeDescriptorSets.emplace_back(perPolyDescSet, 3, 0, vk::DescriptorType::eUniformBuffer, nullptr, lightBufferInfo);
```
Comment (flyinghead) [inline core/rend/vulkan/pipeline.h:110]: uniBufferInfo and lightBufferInfo are used by reference and go out of scope

**Review by flyinghead — CHANGES_REQUESTED** [review]

File: core/rend/vulkan/oit/oit_pipeline.h (hunk 1 above)
Comment (scribam) [reply core/rend/vulkan/oit/oit_pipeline.h:170]: Indeed, fixed

File: core/rend/vulkan/oit/oit_pipeline.h (hunk 2 above)
Comment (scribam) [reply core/rend/vulkan/oit/oit_pipeline.h:177]: Fixed

File: core/rend/vulkan/oit/oit_pipeline.h (hunk 3 above)
Comment (scribam) [reply core/rend/vulkan/oit/oit_pipeline.h:195]: Fixed

File: core/rend/vulkan/pipeline.h (hunk 4 above)
Comment (scribam) [reply core/rend/vulkan/pipeline.h:94]: Fixed

File: core/rend/vulkan/pipeline.h (hunk 5 above)
Comment (scribam) [reply core/rend/vulkan/pipeline.h:112]: Fixed

**Pull request merged**